File: Backtracking/subsets.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
const subsets = (nums) => {
  const results = [];
  const path = [];

  backtrack(results, path, nums, 0);

  return results;
};

const backtrack = (results, path, nums, start) => {
  // add path to results
  results.push([...path]);

  for (let i = start; i < nums.length; i++) {
    // include num
    path.push(nums[i]);
    // backtrack
    backtrack(results, path, nums, i + 1);
    // do not include num
    path.pop();
  }
};

console.log(subsets([1, 2, 3])); // [[],[1],[2],[1,2],[3],[1,3],[2,3],[1,2,3]]
console.log(subsets([0])); // [[],[0]]

File: Backtracking/subsets.py
# LC 78
from typing import List

class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        res = []
        path = []
        
        def backtrack(start):
            res.append(path[:])
            
            for i in range(start, len(nums)):
                path.append(nums[i])   
                backtrack(i + 1)
                path.pop()
        
        backtrack(0)
        return res

solution = Solution()
print(solution.subsets([1,2,3]))